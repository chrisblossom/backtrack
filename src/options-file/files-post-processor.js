/* @flow */

import path from 'path';
import { parseFilePath } from '../utils/parse-file-path';
import { normalizePathname } from '../utils/normalize-pathname';
import { toArray } from '../utils/object-utils';

import type { FileManager, ParsedFiles } from '../types.js';

function shouldSkip(pathname, skipFiles) {
    const split = pathname.split(path.sep);

    let currentPath;
    for (const sub of split) {
        currentPath = currentPath ? currentPath + path.sep + sub : sub;

        if (skipFiles.includes(currentPath)) {
            return true;
        }
    }

    return false;
}

function splitSections(copyFiles) {
    return copyFiles.reduce(
        (acc, file) => {
            if (file.src && file.dest) {
                return {
                    ...acc,
                    copyFiles: [...acc.copyFiles, file],
                };
            }

            // $FlowIssue
            const { makeDirs = [], ...options } = file;

            return {
                ...acc,
                options: [...acc.options, options],
                makeDirs: [...acc.makeDirs, ...makeDirs],
            };
        },
        { copyFiles: [], options: [], makeDirs: [] },
    );
}

function parseOptions(options) {
    return options.reduce(
        (acc, fileOptions) => {
            if (fileOptions.skip) {
                toArray(fileOptions.skip).forEach((relativePath) => {
                    const normalizedFile = normalizePathname(relativePath);

                    acc.skip.push(normalizedFile);
                });
            }

            if (typeof fileOptions.allowChanges === 'boolean') {
                acc.allowChangesAll = fileOptions.allowChanges;

                return acc;
            }

            toArray(fileOptions.allowChanges).forEach((file) => {
                const normalizedFile = normalizePathname(file);

                acc.allowChanges.push(normalizedFile);
            });

            return acc;
        },
        {
            allowChangesAll: false,
            allowChanges: [],
            skip: [],
        },
    );
}

function getMakeDirs(makeDirs, files, skipFiles) {
    const filePaths = files
        .map((filePath) => {
            const { dir } = path.parse(filePath);

            return dir;
        })
        .filter(Boolean);

    const result = [...filePaths, ...makeDirs]
        // eslint-disable-next-line arrow-body-style
        .map((dir) => normalizePathname(dir))
        .sort()
        .reduceRight((acc, dir) => {
            /**
             * remove skipped
             */
            const skip = shouldSkip(dir, skipFiles);

            if (skip === true) {
                return acc;
            }

            /**
             * removes redundant directories
             * ['src', 'dist', 'dist/static'] -> ['src', 'dist/static']
             */
            const previous = acc[0];
            if (previous) {
                const trimmedPrevious = previous.substring(0, dir.length);

                if (trimmedPrevious === dir) {
                    return acc;
                }
            }

            return [dir, ...acc];
        }, []);

    return result;
}

// eslint-disable-next-line flowtype/require-exact-type
type Args = {
    value: FileManager,
};

/**
 * Normally this would be a normal processor, but we need to have all results up front
 * to determine what to exclude etc. I'm sure the function could be rewritten but this works as is.
 */
function filesPostProcessor({ value = [] }: Args = {}): ParsedFiles {
    const result = {
        src: {
            files: [],
            absolute: {},
            hash: {},
        },
        dest: {
            files: [],
            absolute: {},
            hash: {},
            allowChanges: {},
        },
        makeDirs: [],
    };

    const sections = splitSections(value);
    const options = parseOptions(sections.options);

    for (const file of sections.copyFiles) {
        const src = parseFilePath(file.src);
        const dest = parseFilePath(file.dest);

        const skip = shouldSkip(dest.relative, options.skip);
        if (skip === false) {
            const srcId = src.relative;
            result.src.files.push(srcId);
            result.src.absolute[srcId] = src.absolute;
            result.src.hash[srcId] = src.hash;

            const destId = dest.relative;
            result.dest.files.push(destId);
            result.dest.absolute[destId] = dest.absolute;
            result.dest.hash[destId] = dest.hash;

            const allowChanges =
                options.allowChanges.includes(destId) || !!file.allowChanges;

            result.dest.allowChanges[destId] =
                options.allowChangesAll || allowChanges;
        }
    }

    result.makeDirs = getMakeDirs(
        sections.makeDirs,
        result.dest.files,
        options.skip,
    );

    return result;
}

export { filesPostProcessor };
