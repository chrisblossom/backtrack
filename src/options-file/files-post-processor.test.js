/* @flow */

import path from 'path';

const filesPostProcessor = (copyFiles) =>
    require('./files-post-processor').filesPostProcessor(copyFiles);

describe('filesPostProcessor', () => {
    const cwd = process.cwd();

    afterEach(() => {
        process.chdir(cwd);
    });

    test('handles undefined', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const result = filesPostProcessor();

        expect(result).toMatchSnapshot();
    });

    test('returns new object', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [
            {
                src: path.resolve(dir, 'nested/other.js'),
                dest: 'nested/other.js',
            },
            {
                src: path.resolve(dir, 'nested/inside.js'),
                dest: 'nested/inside.js',
            },
            {
                src: path.resolve(dir, 'z-file2.js'),
                dest: 'z-file2.js',
            },
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'file1.js',
            },
        ];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });

    test('returns empty hash when file does not exist', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [
            {
                src: path.resolve(dir, 'nested/missing.js'),
                dest: 'nested/other.js',
            },
            {
                src: path.resolve(dir, 'nested/inside.js'),
                dest: 'nested/missing.js',
            },
            {
                src: path.resolve(dir, 'missing.js'),
                dest: 'missing.js',
            },
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'file1.js',
            },
        ];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });

    test('handles options', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'file1.js',
            },
            {
                src: path.resolve(dir, 'nested/inside.js'),
                dest: 'nested/inside.js',
                allowChanges: true,
            },
            {
                src: path.resolve(dir, 'nested/other.js'),
                dest: 'nested/other.js',
            },
            {
                src: path.resolve(dir, 'z-file2.js'),
                dest: 'z-file2.js',
            },
            {
                skip: ['file1.js'],
                allowChanges: ['z-file2.js'],
            },
        ];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });

    test('handles single options', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'file1.js',
            },
            {
                src: path.resolve(dir, 'nested/inside.js'),
                dest: 'nested/inside.js',
                allowChanges: true,
            },
            {
                src: path.resolve(dir, 'nested/other.js'),
                dest: 'nested/other.js',
            },
            {
                src: path.resolve(dir, 'z-file2.js'),
                dest: 'z-file2.js',
            },
            {
                skip: 'file1.js',
                allowChanges: 'z-file2.js',
            },
        ];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });

    test('handles allowChangesAll', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'file1.js',
            },
            {
                src: path.resolve(dir, 'nested/inside.js'),
                dest: 'nested/inside.js',
            },
            {
                src: path.resolve(dir, 'nested/other.js'),
                dest: 'nested/other.js',
            },
            {
                allowChanges: true,
            },
        ];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });

    test('handles makeDirs', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [
            {
                makeDirs: ['src'],
            },
            {
                makeDirs: [
                    'dist/other/nested',
                    'dist',
                    'dist/static',
                    'static',
                ],
            },
        ];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });

    test('removes duplicate makeDirs', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [{ makeDirs: ['dist', 'src'] }, { makeDirs: ['dist'] }];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });

    test('removes duplicate makeDirs with exact path', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [
            { makeDirs: ['src', 'dist'] },
            { makeDirs: [path.resolve(dir, 'src'), path.resolve(dir, 'dist')] },
        ];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });

    test('makeDirs normalizes paths', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [
            {
                makeDirs: [path.resolve('src/')],
            },
            {
                makeDirs: ['src/'],
            },
            {
                makeDirs: ['src'],
            },
        ];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });

    test('skip works with makeDirs', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [
            {
                makeDirs: ['src', 'dist', 'static'],
            },
            {
                skip: ['src/', 'dist'],
            },
        ];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });

    test('makeDirs skip remove last tree when both are specified', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [
            {
                makeDirs: ['dist', 'dist/static', 'src'],
            },
            {
                skip: ['dist/static'],
            },
        ];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });

    test('makeDirs skip remove whole tree when base specified', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [
            {
                makeDirs: ['dist', 'dist/static', 'src'],
            },
            {
                skip: ['dist'],
            },
        ];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });

    test('skip files if inside skipped dir', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [
            {
                src: path.resolve(dir, 'nested/other.js'),
                dest: 'nested/other.js',
            },
            {
                makeDirs: ['dist', 'dist/static', 'src'],
            },
            {
                skip: ['nested'],
            },
        ];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });

    test('do not skip files if directory does not match', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [
            {
                src: path.resolve(dir, 'nested/inside/other.js'),
                dest: 'nested/inside/other.js',
            },
            {
                makeDirs: ['dist', 'dist/static', 'src'],
            },
            {
                skip: ['nested/outside'],
            },
        ];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });

    test('adds file paths to makeDirs', () => {
        const dir = path.resolve(
            __dirname,
            '../file-manager/__sandbox__/stats1/',
        );
        process.chdir(dir);

        const value = [
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'dist/file2.js',
            },
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'file1.js',
            },
            {
                src: path.resolve(dir, 'nested/inside/other.js'),
                dest: 'nested/inside/other.js',
            },
            {
                src: path.resolve(dir, 'nested/outside/other.js'),
                dest: 'nested/outside/other.js',
            },
            {
                makeDirs: ['dist', 'dist/static', 'src'],
                skip: ['nested/outside'],
            },
        ];

        const result = filesPostProcessor({ value });

        expect(result).toMatchSnapshot();
    });
});
