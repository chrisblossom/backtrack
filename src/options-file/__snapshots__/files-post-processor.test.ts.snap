// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filesPostProcessor adds file paths to makeDirs 1`] = `
{
  "dest": {
    "absolute": {
      "dist/file2.js": "<PROJECT_ROOT>/dist/file2.js",
      "file1.js": "<PROJECT_ROOT>/file1.js",
      "nested/inside/other.js": "<PROJECT_ROOT>/nested/inside/other.js",
    },
    "allowChanges": {
      "dist/file2.js": false,
      "file1.js": false,
      "nested/inside/other.js": false,
    },
    "files": [
      "dist/file2.js",
      "file1.js",
      "nested/inside/other.js",
    ],
    "hash": {
      "dist/file2.js": "",
      "file1.js": "971eb09e78cb5dca6dd4fba0284b996b624df850e0e27d2933280271d3a7b856",
      "nested/inside/other.js": "",
    },
    "ignoreUpdates": {
      "dist/file2.js": false,
      "file1.js": false,
      "nested/inside/other.js": false,
    },
  },
  "makeDirs": [
    "dist/static",
    "nested/inside",
    "src",
  ],
  "src": {
    "absolute": {
      "file1.js": "<PROJECT_ROOT>/file1.js",
      "nested/inside/other.js": "<PROJECT_ROOT>/nested/inside/other.js",
    },
    "files": [
      "file1.js",
      "file1.js",
      "nested/inside/other.js",
    ],
    "hash": {
      "file1.js": "971eb09e78cb5dca6dd4fba0284b996b624df850e0e27d2933280271d3a7b856",
      "nested/inside/other.js": "",
    },
  },
}
`;

exports[`filesPostProcessor do not skip files if directory does not match 1`] = `
{
  "dest": {
    "absolute": {
      "nested/inside/other.js": "<PROJECT_ROOT>/nested/inside/other.js",
    },
    "allowChanges": {
      "nested/inside/other.js": false,
    },
    "files": [
      "nested/inside/other.js",
    ],
    "hash": {
      "nested/inside/other.js": "",
    },
    "ignoreUpdates": {
      "nested/inside/other.js": false,
    },
  },
  "makeDirs": [
    "dist/static",
    "nested/inside",
    "src",
  ],
  "src": {
    "absolute": {
      "nested/inside/other.js": "<PROJECT_ROOT>/nested/inside/other.js",
    },
    "files": [
      "nested/inside/other.js",
    ],
    "hash": {
      "nested/inside/other.js": "",
    },
  },
}
`;

exports[`filesPostProcessor handles allowChangesAll 1`] = `
{
  "dest": {
    "absolute": {
      "file1.js": "<PROJECT_ROOT>/file1.js",
      "nested/inside.js": "<PROJECT_ROOT>/nested/inside.js",
      "nested/other.js": "<PROJECT_ROOT>/nested/other.js",
    },
    "allowChanges": {
      "file1.js": true,
      "nested/inside.js": true,
      "nested/other.js": true,
    },
    "files": [
      "file1.js",
      "nested/inside.js",
      "nested/other.js",
    ],
    "hash": {
      "file1.js": "971eb09e78cb5dca6dd4fba0284b996b624df850e0e27d2933280271d3a7b856",
      "nested/inside.js": "46234eada9c51bae5d7a671fbb939e6efd1dda8b524f2b8b015aa4bce7746c29",
      "nested/other.js": "44c09b273fa7e0ffcb87b76cf882a5c7cdeba79e80b8bc814d3909a2a58ad995",
    },
    "ignoreUpdates": {
      "file1.js": false,
      "nested/inside.js": false,
      "nested/other.js": false,
    },
  },
  "makeDirs": [
    "nested",
  ],
  "src": {
    "absolute": {
      "file1.js": "<PROJECT_ROOT>/file1.js",
      "nested/inside.js": "<PROJECT_ROOT>/nested/inside.js",
      "nested/other.js": "<PROJECT_ROOT>/nested/other.js",
    },
    "files": [
      "file1.js",
      "nested/inside.js",
      "nested/other.js",
    ],
    "hash": {
      "file1.js": "971eb09e78cb5dca6dd4fba0284b996b624df850e0e27d2933280271d3a7b856",
      "nested/inside.js": "46234eada9c51bae5d7a671fbb939e6efd1dda8b524f2b8b015aa4bce7746c29",
      "nested/other.js": "44c09b273fa7e0ffcb87b76cf882a5c7cdeba79e80b8bc814d3909a2a58ad995",
    },
  },
}
`;

exports[`filesPostProcessor handles ignoreUpdates 1`] = `
{
  "dest": {
    "absolute": {
      "file1.js": "<PROJECT_ROOT>/file1.js",
    },
    "allowChanges": {
      "file1.js": true,
    },
    "files": [
      "file1.js",
    ],
    "hash": {
      "file1.js": "971eb09e78cb5dca6dd4fba0284b996b624df850e0e27d2933280271d3a7b856",
    },
    "ignoreUpdates": {
      "file1.js": true,
    },
  },
  "makeDirs": [],
  "src": {
    "absolute": {
      "file1.js": "<PROJECT_ROOT>/file1.js",
    },
    "files": [
      "file1.js",
    ],
    "hash": {
      "file1.js": "ignoreUpdates",
    },
  },
}
`;

exports[`filesPostProcessor handles ignoreUpdatesAll 1`] = `
{
  "dest": {
    "absolute": {
      "file1.js": "<PROJECT_ROOT>/file1.js",
      "nested/inside.js": "<PROJECT_ROOT>/nested/inside.js",
      "nested/other.js": "<PROJECT_ROOT>/nested/other.js",
    },
    "allowChanges": {
      "file1.js": true,
      "nested/inside.js": true,
      "nested/other.js": true,
    },
    "files": [
      "file1.js",
      "nested/inside.js",
      "nested/other.js",
    ],
    "hash": {
      "file1.js": "971eb09e78cb5dca6dd4fba0284b996b624df850e0e27d2933280271d3a7b856",
      "nested/inside.js": "46234eada9c51bae5d7a671fbb939e6efd1dda8b524f2b8b015aa4bce7746c29",
      "nested/other.js": "44c09b273fa7e0ffcb87b76cf882a5c7cdeba79e80b8bc814d3909a2a58ad995",
    },
    "ignoreUpdates": {
      "file1.js": true,
      "nested/inside.js": true,
      "nested/other.js": true,
    },
  },
  "makeDirs": [
    "nested",
  ],
  "src": {
    "absolute": {
      "file1.js": "<PROJECT_ROOT>/file1.js",
      "nested/inside.js": "<PROJECT_ROOT>/nested/inside.js",
      "nested/other.js": "<PROJECT_ROOT>/nested/other.js",
    },
    "files": [
      "file1.js",
      "nested/inside.js",
      "nested/other.js",
    ],
    "hash": {
      "file1.js": "ignoreUpdates",
      "nested/inside.js": "ignoreUpdates",
      "nested/other.js": "ignoreUpdates",
    },
  },
}
`;

exports[`filesPostProcessor handles makeDirs 1`] = `
{
  "dest": {
    "absolute": {},
    "allowChanges": {},
    "files": [],
    "hash": {},
    "ignoreUpdates": {},
  },
  "makeDirs": [
    "dist/other/nested",
    "dist/static",
    "src",
    "static",
  ],
  "src": {
    "absolute": {},
    "files": [],
    "hash": {},
  },
}
`;

exports[`filesPostProcessor handles options 1`] = `
{
  "dest": {
    "absolute": {
      "nested/inside.js": "<PROJECT_ROOT>/nested/inside.js",
      "nested/other.js": "<PROJECT_ROOT>/nested/other.js",
      "z-file2.js": "<PROJECT_ROOT>/z-file2.js",
    },
    "allowChanges": {
      "nested/inside.js": true,
      "nested/other.js": false,
      "z-file2.js": true,
    },
    "files": [
      "nested/inside.js",
      "nested/other.js",
      "z-file2.js",
    ],
    "hash": {
      "nested/inside.js": "46234eada9c51bae5d7a671fbb939e6efd1dda8b524f2b8b015aa4bce7746c29",
      "nested/other.js": "44c09b273fa7e0ffcb87b76cf882a5c7cdeba79e80b8bc814d3909a2a58ad995",
      "z-file2.js": "447ad688b0d3b7449000199a9e766ee03965da1ab15a961756a90dcff3a810e3",
    },
    "ignoreUpdates": {
      "nested/inside.js": false,
      "nested/other.js": false,
      "z-file2.js": false,
    },
  },
  "makeDirs": [
    "nested",
  ],
  "src": {
    "absolute": {
      "nested/inside.js": "<PROJECT_ROOT>/nested/inside.js",
      "nested/other.js": "<PROJECT_ROOT>/nested/other.js",
      "z-file2.js": "<PROJECT_ROOT>/z-file2.js",
    },
    "files": [
      "nested/inside.js",
      "nested/other.js",
      "z-file2.js",
    ],
    "hash": {
      "nested/inside.js": "46234eada9c51bae5d7a671fbb939e6efd1dda8b524f2b8b015aa4bce7746c29",
      "nested/other.js": "44c09b273fa7e0ffcb87b76cf882a5c7cdeba79e80b8bc814d3909a2a58ad995",
      "z-file2.js": "447ad688b0d3b7449000199a9e766ee03965da1ab15a961756a90dcff3a810e3",
    },
  },
}
`;

exports[`filesPostProcessor handles single options 1`] = `
{
  "dest": {
    "absolute": {
      "ignore-updates-1.js": "<PROJECT_ROOT>/ignore-updates-1.js",
      "ignore-updates-2.js": "<PROJECT_ROOT>/ignore-updates-2.js",
      "nested/inside.js": "<PROJECT_ROOT>/nested/inside.js",
      "nested/other.js": "<PROJECT_ROOT>/nested/other.js",
      "z-file2.js": "<PROJECT_ROOT>/z-file2.js",
    },
    "allowChanges": {
      "ignore-updates-1.js": true,
      "ignore-updates-2.js": true,
      "nested/inside.js": true,
      "nested/other.js": false,
      "z-file2.js": true,
    },
    "files": [
      "nested/inside.js",
      "nested/other.js",
      "ignore-updates-1.js",
      "ignore-updates-2.js",
      "z-file2.js",
    ],
    "hash": {
      "ignore-updates-1.js": "",
      "ignore-updates-2.js": "",
      "nested/inside.js": "46234eada9c51bae5d7a671fbb939e6efd1dda8b524f2b8b015aa4bce7746c29",
      "nested/other.js": "44c09b273fa7e0ffcb87b76cf882a5c7cdeba79e80b8bc814d3909a2a58ad995",
      "z-file2.js": "447ad688b0d3b7449000199a9e766ee03965da1ab15a961756a90dcff3a810e3",
    },
    "ignoreUpdates": {
      "ignore-updates-1.js": true,
      "ignore-updates-2.js": true,
      "nested/inside.js": false,
      "nested/other.js": false,
      "z-file2.js": false,
    },
  },
  "makeDirs": [
    "nested",
  ],
  "src": {
    "absolute": {
      "nested/ignore-updates-1.js": "<PROJECT_ROOT>/nested/ignore-updates-1.js",
      "nested/ignore-updates-2.js": "<PROJECT_ROOT>/nested/ignore-updates-2.js",
      "nested/inside.js": "<PROJECT_ROOT>/nested/inside.js",
      "nested/other.js": "<PROJECT_ROOT>/nested/other.js",
      "z-file2.js": "<PROJECT_ROOT>/z-file2.js",
    },
    "files": [
      "nested/inside.js",
      "nested/other.js",
      "nested/ignore-updates-1.js",
      "nested/ignore-updates-2.js",
      "z-file2.js",
    ],
    "hash": {
      "nested/ignore-updates-1.js": "ignoreUpdates",
      "nested/ignore-updates-2.js": "ignoreUpdates",
      "nested/inside.js": "46234eada9c51bae5d7a671fbb939e6efd1dda8b524f2b8b015aa4bce7746c29",
      "nested/other.js": "44c09b273fa7e0ffcb87b76cf882a5c7cdeba79e80b8bc814d3909a2a58ad995",
      "z-file2.js": "447ad688b0d3b7449000199a9e766ee03965da1ab15a961756a90dcff3a810e3",
    },
  },
}
`;

exports[`filesPostProcessor handles undefined 1`] = `
{
  "dest": {
    "absolute": {},
    "allowChanges": {},
    "files": [],
    "hash": {},
    "ignoreUpdates": {},
  },
  "makeDirs": [],
  "src": {
    "absolute": {},
    "files": [],
    "hash": {},
  },
}
`;

exports[`filesPostProcessor makeDirs normalizes paths 1`] = `
{
  "dest": {
    "absolute": {},
    "allowChanges": {},
    "files": [],
    "hash": {},
    "ignoreUpdates": {},
  },
  "makeDirs": [
    "src",
  ],
  "src": {
    "absolute": {},
    "files": [],
    "hash": {},
  },
}
`;

exports[`filesPostProcessor makeDirs skip remove last tree when both are specified 1`] = `
{
  "dest": {
    "absolute": {},
    "allowChanges": {},
    "files": [],
    "hash": {},
    "ignoreUpdates": {},
  },
  "makeDirs": [
    "dist",
    "src",
  ],
  "src": {
    "absolute": {},
    "files": [],
    "hash": {},
  },
}
`;

exports[`filesPostProcessor makeDirs skip remove whole tree when base specified 1`] = `
{
  "dest": {
    "absolute": {},
    "allowChanges": {},
    "files": [],
    "hash": {},
    "ignoreUpdates": {},
  },
  "makeDirs": [
    "src",
  ],
  "src": {
    "absolute": {},
    "files": [],
    "hash": {},
  },
}
`;

exports[`filesPostProcessor removes duplicate dest 1`] = `
{
  "dest": {
    "absolute": {
      "file1.js": "<PROJECT_ROOT>/file1.js",
      "nested/other.js": "<PROJECT_ROOT>/nested/other.js",
    },
    "allowChanges": {
      "file1.js": true,
      "nested/other.js": false,
    },
    "files": [
      "file1.js",
      "nested/other.js",
    ],
    "hash": {
      "file1.js": "971eb09e78cb5dca6dd4fba0284b996b624df850e0e27d2933280271d3a7b856",
      "nested/other.js": "44c09b273fa7e0ffcb87b76cf882a5c7cdeba79e80b8bc814d3909a2a58ad995",
    },
    "ignoreUpdates": {
      "file1.js": true,
      "nested/other.js": false,
    },
  },
  "makeDirs": [
    "fake/dir",
    "nested",
  ],
  "src": {
    "absolute": {
      "nested/other.js": "<PROJECT_ROOT>/nested/other.js",
      "z-file2.js": "<PROJECT_ROOT>/z-file2.js",
    },
    "files": [
      "z-file2.js",
      "nested/other.js",
    ],
    "hash": {
      "nested/other.js": "44c09b273fa7e0ffcb87b76cf882a5c7cdeba79e80b8bc814d3909a2a58ad995",
      "z-file2.js": "ignoreUpdates",
    },
  },
}
`;

exports[`filesPostProcessor removes duplicate makeDirs 1`] = `
{
  "dest": {
    "absolute": {},
    "allowChanges": {},
    "files": [],
    "hash": {},
    "ignoreUpdates": {},
  },
  "makeDirs": [
    "dist",
    "src",
  ],
  "src": {
    "absolute": {},
    "files": [],
    "hash": {},
  },
}
`;

exports[`filesPostProcessor removes duplicate makeDirs with exact path 1`] = `
{
  "dest": {
    "absolute": {},
    "allowChanges": {},
    "files": [],
    "hash": {},
    "ignoreUpdates": {},
  },
  "makeDirs": [
    "dist",
    "src",
  ],
  "src": {
    "absolute": {},
    "files": [],
    "hash": {},
  },
}
`;

exports[`filesPostProcessor returns empty hash when file does not exist 1`] = `
{
  "dest": {
    "absolute": {
      "file1.js": "<PROJECT_ROOT>/file1.js",
      "missing.js": "<PROJECT_ROOT>/missing.js",
      "nested/missing.js": "<PROJECT_ROOT>/nested/missing.js",
      "nested/other.js": "<PROJECT_ROOT>/nested/other.js",
    },
    "allowChanges": {
      "file1.js": false,
      "missing.js": false,
      "nested/missing.js": false,
      "nested/other.js": false,
    },
    "files": [
      "nested/other.js",
      "nested/missing.js",
      "missing.js",
      "file1.js",
    ],
    "hash": {
      "file1.js": "971eb09e78cb5dca6dd4fba0284b996b624df850e0e27d2933280271d3a7b856",
      "missing.js": "",
      "nested/missing.js": "",
      "nested/other.js": "44c09b273fa7e0ffcb87b76cf882a5c7cdeba79e80b8bc814d3909a2a58ad995",
    },
    "ignoreUpdates": {
      "file1.js": false,
      "missing.js": false,
      "nested/missing.js": false,
      "nested/other.js": false,
    },
  },
  "makeDirs": [
    "nested",
  ],
  "src": {
    "absolute": {
      "file1.js": "<PROJECT_ROOT>/file1.js",
      "missing.js": "<PROJECT_ROOT>/missing.js",
      "nested/inside.js": "<PROJECT_ROOT>/nested/inside.js",
      "nested/missing.js": "<PROJECT_ROOT>/nested/missing.js",
    },
    "files": [
      "nested/missing.js",
      "nested/inside.js",
      "missing.js",
      "file1.js",
    ],
    "hash": {
      "file1.js": "971eb09e78cb5dca6dd4fba0284b996b624df850e0e27d2933280271d3a7b856",
      "missing.js": "",
      "nested/inside.js": "46234eada9c51bae5d7a671fbb939e6efd1dda8b524f2b8b015aa4bce7746c29",
      "nested/missing.js": "",
    },
  },
}
`;

exports[`filesPostProcessor returns new object 1`] = `
{
  "dest": {
    "absolute": {
      "file1.js": "<PROJECT_ROOT>/file1.js",
      "nested/inside.js": "<PROJECT_ROOT>/nested/inside.js",
      "nested/other.js": "<PROJECT_ROOT>/nested/other.js",
      "z-file2.js": "<PROJECT_ROOT>/z-file2.js",
    },
    "allowChanges": {
      "file1.js": false,
      "nested/inside.js": false,
      "nested/other.js": false,
      "z-file2.js": false,
    },
    "files": [
      "nested/other.js",
      "nested/inside.js",
      "z-file2.js",
      "file1.js",
    ],
    "hash": {
      "file1.js": "971eb09e78cb5dca6dd4fba0284b996b624df850e0e27d2933280271d3a7b856",
      "nested/inside.js": "46234eada9c51bae5d7a671fbb939e6efd1dda8b524f2b8b015aa4bce7746c29",
      "nested/other.js": "44c09b273fa7e0ffcb87b76cf882a5c7cdeba79e80b8bc814d3909a2a58ad995",
      "z-file2.js": "447ad688b0d3b7449000199a9e766ee03965da1ab15a961756a90dcff3a810e3",
    },
    "ignoreUpdates": {
      "file1.js": false,
      "nested/inside.js": false,
      "nested/other.js": false,
      "z-file2.js": false,
    },
  },
  "makeDirs": [
    "nested",
  ],
  "src": {
    "absolute": {
      "file1.js": "<PROJECT_ROOT>/file1.js",
      "nested/inside.js": "<PROJECT_ROOT>/nested/inside.js",
      "nested/other.js": "<PROJECT_ROOT>/nested/other.js",
      "z-file2.js": "<PROJECT_ROOT>/z-file2.js",
    },
    "files": [
      "nested/other.js",
      "nested/inside.js",
      "z-file2.js",
      "file1.js",
    ],
    "hash": {
      "file1.js": "971eb09e78cb5dca6dd4fba0284b996b624df850e0e27d2933280271d3a7b856",
      "nested/inside.js": "46234eada9c51bae5d7a671fbb939e6efd1dda8b524f2b8b015aa4bce7746c29",
      "nested/other.js": "44c09b273fa7e0ffcb87b76cf882a5c7cdeba79e80b8bc814d3909a2a58ad995",
      "z-file2.js": "447ad688b0d3b7449000199a9e766ee03965da1ab15a961756a90dcff3a810e3",
    },
  },
}
`;

exports[`filesPostProcessor skip files if inside skipped dir 1`] = `
{
  "dest": {
    "absolute": {},
    "allowChanges": {},
    "files": [],
    "hash": {},
    "ignoreUpdates": {},
  },
  "makeDirs": [
    "dist/static",
    "src",
  ],
  "src": {
    "absolute": {},
    "files": [],
    "hash": {},
  },
}
`;

exports[`filesPostProcessor skip works with makeDirs 1`] = `
{
  "dest": {
    "absolute": {},
    "allowChanges": {},
    "files": [],
    "hash": {},
    "ignoreUpdates": {},
  },
  "makeDirs": [
    "static",
  ],
  "src": {
    "absolute": {},
    "files": [],
    "hash": {},
  },
}
`;
